src/products/service/product.service.ts

```diff
@@ -0,0 +1,176 @@
+            );
+        }
+
+        const newProduct = new this.productModel({
+            ...productDetails,
+            store: storeID,
+        });
+
+        newProduct.save();
+
+        return {
+            message: "product created succesfully"
+        }
+    }
+
+    async getAllProducts() {
+        const products = await this.productModel.find();
+        return products
+    }
+
+    async getProductById(productId: string) {
+        const product = await this.productModel.findOne({ _id: productId, })
+
+        if (!product) {
+            throw new AppError(
+                `Product not found`,
+                HttpStatus.NOT_FOUND,
+            );
+        }
+
+        return product
+    }
+
+    async findProductsByStoreIDForRetailer(userId: string, storeId: string) {
+        const store = await this.storeModel.findOne({
+            _id: storeId,
+            owner: userId,
+        });
+
+        if (!store) {
+            throw new AppError(
+                `Store not found`,
+                HttpStatus.NOT_FOUND,
+            );
+        }
+        const products = await this.productModel.find({
+            store: storeId,
+        });
+
+        return products;
+    }
+
+    async updateMyProduct(
+        userId: string,
+        updateProductDto: UpdateProductDTO
+    ) {
+
+        const { productId } = updateProductDto;
+
+        const product = await this.productModel.findOne({
+            _id: productId,
+        })
+            .populate({
+                path: 'store',
+                select: 'owner',
+            })
+            .lean();
+
+        if (!product) {
+            throw new AppError(
+                `Product not found`,
+                HttpStatus.NOT_FOUND,
+            );
+        }
+
+        const store = product.store as unknown as { owner: Types.ObjectId };
+
+        if (store.owner.toString() !== userId) {
+            throw new AppError(
+                `Forbidden`,
+                HttpStatus.FORBIDDEN,
+            );
+        }
+
+        await this.productModel.findByIdAndUpdate(
+            productId,
+            updateProductDto,
+            { new: true }
+        );
+
+        return { message: 'Product updated succesfully' };
+    }
+
+    async deleteMyProduct(userId: string, productId: string) {
+        const product = await this.productModel.findOne({
+            _id: productId,
+        }).populate({
+            path: 'store',
+            select: 'owner',
+        }).lean();
+
+        if (!product) {
+            throw new AppError(
+                `Product not found`,
+                HttpStatus.NOT_FOUND,
+            );
+        }
+
+        const store = product.store as unknown as { owner: Types.ObjectId };
+
+        const isStoreOwner = store.owner.toString() === userId;
+
+        if (!isStoreOwner) {
+            throw new AppError(
+                `Forbidden`,
+                HttpStatus.FORBIDDEN,
+            );
+        }
```

Comment:
Why didn't you just pass the userId into the query so when you don't find the product, you can as well just throw an error and move on

Reply:
that way , for some reasons I don't know , it returns "null" for the product

Reply:
const product = await this.productModel.findOne({
            _id: productId,
            
            store: {
              _id: userId,
            },
            
          });